### DISCUSSION TITLE
LevelDB 源码分析「三、高性能写操作」 | SF-Zhou's Blog

### DISCUSSION THREAD

**Discussion by SF-Zhou** [SF-Zhou/sf-zhou.github.io#173] [Announcements]
https://sf-zhou.github.io/leveldb/leveldb_03_write_batch_and_log.html

**Comment by suntzu93** [comment]
Thank you for the great article. I'm adding trying to use encrypt/decrypt for leveldb and my idea is encrypt data before write to file in env_posix.cc , method WriteUnbuffered(const char *data, size_t size) . But I don't know if encrypt " char *data " has any effect to Record Header , CRC32... . I greatly appreciate your comments on this matter.

**Comment by SF-Zhou** [comment]
> @suntzu93
> Thank you for the great article. I'm adding trying to use encrypt/decrypt for leveldb and my idea is encrypt data before write to file in env_posix.cc , method WriteUnbuffered(const char *data, size_t size) . But I don't know if encrypt " char *data " has any effect to Record Header , CRC32... . I greatly appreciate your comments on this matter. 

I think it's ok if you do symmetrical decryption after `PosixRandomAccessFile::Read`. The position of reading and writing may need to be mapped according to the encryption algorithm.

**Comment by suntzu93** [comment]
> > @suntzu93
> > Thank you for the great article. I'm adding trying to use encrypt/decrypt for leveldb and my idea is encrypt data before write to file in env_posix.cc , method WriteUnbuffered(const char *data, size_t size) . But I don't know if encrypt " char *data " has any effect to Record Header , CRC32... . I greatly appreciate your comments on this matter.
> 
> I think it's ok if you do symmetrical decryption after `PosixRandomAccessFile::Read`. The position of reading and writing may need to be mapped according to the encryption algorithm.

Thanks for your answer, I'm doing with AES. I created a dex array from *data but only the first 5 bytes can be encrypted, the key and value cannot be encrypted. I'm using this library https://github.com/kokke/tiny-AES-c and it's working perfect with normal string. Do you have any idea about the type of  "char *data"?

**Comment by SF-Zhou** [comment]
> @suntzu93
> > > @suntzu93
> > > Thank you for the great article. I'm adding trying to use encrypt/decrypt for leveldb and my idea is encrypt data before write to file in env_posix.cc , method WriteUnbuffered(const char *data, size_t size) . But I don't know if encrypt " char *data " has any effect to Record Header , CRC32... . I greatly appreciate your comments on this matter.
> > 
> > I think it's ok if you do symmetrical decryption after `PosixRandomAccessFile::Read`. The position of reading and writing may need to be mapped according to the encryption algorithm.
> 
> Thanks for your answer, I'm doing with AES. I created a dex array from *data but only the first 5 bytes can be encrypted, the key and value cannot be encrypted. I'm using this library https://github.com/kokke/tiny-AES-c and it's working perfect with normal string. Do you have any idea about the type of  "char *data"?

The `data` in `WriteUnbuffered(const char *data, size_t size)` is not a c string, but a byte stream. It can contain `\0`. For the AES algorithm, I have not used it before and may not be able to help.

**Comment by suntzu93** [comment]
Really thank you for your help, now I have a keywork to continue my work. Have a nice day bro.

**Comment by Stephan14** [comment]
看了一下代码，写入的时候好像是按照Record粒度来写入文件的，不是按照Block粒度来写入文件的，对吗？

**Reply by Stephan14** [reply]
当Record很小的时候，是按照Record来写入的；当Record很大的时候，是按照Block粒度来写入的，是这样的吗？

**Reply by SF-Zhou** [reply]
是的
